# **Fetch Assessment - Evaluating Data Quality Issues in the dataset**

## **Findings Summary**:
### **Null Values:**

* There are many null value columns present in all three datasets.
* Some columns have over 50% null values, especially in the "brands" table, which could complicate preprocessing and analysis.
* **Recommendation**: Decide on strategies to handle null values, such as imputation, removal of columns, or other techniques based on the analysis goals.

### **Duplicates:**

* The "users" table contains over 250 duplicate rows.
* **Recommendation**: Remove duplicate rows from the "users" table to ensure data integrity.

### **Receipt Table Issues:**

* Upon manual inspection of the "receipt" table, additional data quality issues were identified.
* A major issue is the absence of a column related to brands, which complicates joining with the "brands" table.
* Another issue found is the presence of duplicate receipt items within a single receipt.
* **Recommendation**:If possible, add a column related to brands in the "receipt" table for easier integration with the "brands" table and address the issue of duplicate receipt items, possibly by investigating their source and deciding on appropriate actions.

### **Further Analysis:**

Consider further exploratory data analysis to gain insights into relationships between tables and to understand the impact of data quality issues on analysis results.

### Users Dataset

In [1]:
# loading brands dataset
import json
import pandas as pd

users = []
with open('users.json', 'r') as file:
    for line in file:
        users.append(json.loads(line))

user_df = pd.DataFrame(users)

In [2]:
# cheking columns with dictionary values
for column in user_df.columns:
    if user_df[column].apply(lambda x: isinstance(x, dict)).any():
        print(f"Column '{column}' contains dictionary values.")

Column '_id' contains dictionary values.
Column 'createdDate' contains dictionary values.
Column 'lastLogin' contains dictionary values.


In [3]:
# converting them to string for evaluating
user_df["_id"] = user_df["_id"].apply(lambda x: str(x))
user_df["createdDate"] = user_df["createdDate"].apply(lambda x: str(x))
user_df["lastLogin"] = user_df["lastLogin"].apply(lambda x: str(x))

In [4]:
user_df.head()

_id  active               createdDate  \
0  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
1  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
2  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
3  {'$oid': '5ff1e1eacfcf6c399c274ae6'}    True  {'$date': 1609687530554}   
4  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   

                  lastLogin      role signUpSource state  
0  {'$date': 1609687537858}  consumer        Email    WI  
1  {'$date': 1609687537858}  consumer        Email    WI  
2  {'$date': 1609687537858}  consumer        Email    WI  
3  {'$date': 1609687530597}  consumer        Email    WI  
4  {'$date': 1609687537858}  consumer        Email    WI

In [5]:
# Check for duplicate rows
duplicate_rows = user_df[user_df.duplicated()]
print(f"Number of duplicate rows: {len(duplicate_rows)}")

# Check for missing values
null_percent = round((user_df.isna().sum() / user_df.shape[0]) * 100, 2)
print(null_percent)

Number of duplicate rows: 283
_id              0.00
active           0.00
createdDate      0.00
lastLogin        0.00
role             0.00
signUpSource     9.70
state           11.31
dtype: float64


*   **From the above, we can see that state has around 11% and signupsource column has around 10% of null values present in them**
*   **Also we can see around 283 duplicate values in this dataset**

### Brands Dataset

In [6]:
# loading brands dataset
brands = []
with open('brands.json', 'r') as file:
    for line in file:
        brands.append(json.loads(line))

brands_df = pd.DataFrame(brands)

In [13]:
# cheking columns with dictionary values
for column in brands_df.columns:
    if brands_df[column].apply(lambda x: isinstance(x, dict)).any():
        print(f"Column '{column}' contains dictionary values.")

Column '_id' contains dictionary values.
Column 'cpg' contains dictionary values.


In [14]:
# converting them to string for evaluating
brands_df["_id"] = brands_df["_id"].apply(lambda x: str(x))
brands_df["cpg"] = brands_df["cpg"].apply(lambda x: str(x))

In [15]:
brands_df.head()

_id       barcode        category  \
0  {'$oid': '601ac115be37ce2ead437551'}  511111019862          Baking   
1  {'$oid': '601c5460be37ce2ead43755f'}  511111519928       Beverages   
2  {'$oid': '601ac142be37ce2ead43755d'}  511111819905          Baking   
3  {'$oid': '601ac142be37ce2ead43755a'}  511111519874          Baking   
4  {'$oid': '601ac142be37ce2ead43755e'}  511111319917  Candy & Sweets   

       categoryCode                                                cpg  \
0            BAKING  {'$id': {'$oid': '601ac114be37ce2ead437550'}, ...   
1         BEVERAGES  {'$id': {'$oid': '5332f5fbe4b03c9a25efd0ba'}, ...   
2            BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
3            BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
4  CANDY_AND_SWEETS  {'$id': {'$oid': '5332fa12e4b03c9a25efd1e7'}, ...   

                        name topBrand                      brandCode  
0  test brand @1612366101024    False                            NaN  
1                  Starbucks    False                      STARBUCKS  
2  test brand @1612366146176    False  TEST BRANDCODE @1612366146176  
3  test brand @1612366146051    False  TEST BRANDCODE @1612366146051  
4  test brand @1612366146827    False  TEST BRANDCODE @1612366146827

In [16]:
# Check for duplicate rows
duplicate_rows = brands_df[brands_df.duplicated()]
print(f"Number of duplicate rows: {len(duplicate_rows)}")

# Check for missing values
null_percent = round((brands_df.isna().sum() / brands_df.shape[0]) * 100, 2)
print(null_percent)

Number of duplicate rows: 0
_id              0.00
barcode          0.00
category        13.28
categoryCode    55.70
cpg              0.00
name             0.00
topBrand        52.44
brandCode       20.05
dtype: float64


**From the above, we can see that:**
*   **TopBrand and categoryCode have over 50% nulls values**
*   **While brandCode and category have 20% and 13% null values respectively**


### Receipts Dataset

In [28]:
# loading brands dataset
receipts = []
with open('receipts.json', 'r') as file:
    for line in file:
        receipts.append(json.loads(line))

receipts_df = pd.DataFrame(receipts)

In [30]:
# exploading the dictionary
receipts_df = receipts_df.explode('rewardsReceiptItemList')
receipts_df.reset_index(inplace=True)

In [31]:
# cheking columns with dictionary values
for column in receipts_df.columns:
    if receipts_df[column].apply(lambda x: isinstance(x, dict)).any():
        print(f"Column '{column}' contains dictionary values.")

Column '_id' contains dictionary values.
Column 'createDate' contains dictionary values.
Column 'dateScanned' contains dictionary values.
Column 'finishedDate' contains dictionary values.
Column 'modifyDate' contains dictionary values.
Column 'pointsAwardedDate' contains dictionary values.
Column 'purchaseDate' contains dictionary values.
Column 'rewardsReceiptItemList' contains dictionary values.


In [32]:
# converting them to string for evaluating
receipts_df["_id"] = receipts_df["_id"].apply(lambda x: str(x))
receipts_df["createDate"] = receipts_df["createDate"].apply(lambda x: str(x))
receipts_df["dateScanned"] = receipts_df["dateScanned"].apply(lambda x: str(x))
receipts_df["finishedDate"] = receipts_df["finishedDate"].apply(lambda x: str(x))
receipts_df["modifyDate"] = receipts_df["modifyDate"].apply(lambda x: str(x))
receipts_df["pointsAwardedDate"] = receipts_df["pointsAwardedDate"].apply(lambda x: str(x))
receipts_df["purchaseDate"] = receipts_df["purchaseDate"].apply(lambda x: str(x))
receipts_df["rewardsReceiptItemList"] = receipts_df["rewardsReceiptItemList"].apply(lambda x: str(x))

In [33]:
# Check for duplicate rows
duplicate_rows = receipts_df[receipts_df.duplicated()]
print(f"Number of duplicate rows: {len(duplicate_rows)}")

# Check for missing values
null_percent = round((receipts_df.isna().sum() / receipts_df.shape[0]) * 100, 2)
print(null_percent)

Number of duplicate rows: 0
index                       0.00
_id                         0.00
bonusPointsEarned          18.98
bonusPointsEarnedReason    18.98
createDate                  0.00
dateScanned                 0.00
finishedDate                0.00
modifyDate                  0.00
pointsAwardedDate           0.00
pointsEarned               15.28
purchaseDate                0.00
purchasedItemCount          6.56
rewardsReceiptItemList      0.00
rewardsReceiptStatus        0.00
totalSpent                  5.89
userId                      0.00
dtype: float64


**From the above we can see that:**
* **bonusPointsEarned and bonusPointsEarnedReason have around 19% nulls**
* **pointsEarned and totalSpent have 15 and 6% nulls respectively**
* **There were no duplicate rows found in the data**

### **Note:**
Apart from this there were many other data quality issues found in the receipts datasets, mainly under the rewardsReceiptItemList. Some of them are:
* No brand related columns present in the receipt which can be used to join the items to a specific product in brands table
* Duplicate receipt items are present under same receipt, instead of increasing the quantityPurchased value.
* Data inconsistencies found, many columns are missing in the dataset
* Barcode present in receipts table is not same as the barcode present in brands table